chemprop/models/ffn.py

```diff
@@ -83,27 +114,36 @@ def __init__(self,
                  output_size: int,
                  dropout: nn.Module,
                  activation: nn.Module,
+                 ffn_base: nn.Module,
                  constraint: bool = False,
-                 ffn_type: str = 'atom'):
+                 ffn_type: str = 'atom',
+                 shared_ffn: bool = True):
         """
         :param features_size: Dimensionality of input features.
         :param hidden_size: Dimensionality of hidden layers.
         :param num_layers: Number of layers in FFN.
         :param output_size: The size of output.
         :param dropout: Dropout probability.
         :param activation: Activation function.
+        :param ffn_base: The shared base layers (all but the last) of the FFN between tasks.
         :param constraint: Whether to apply constraint to output.
         :param ffn_type: The type of target (atom or bond).
+        :param shared_ffn: Whether to share weights in the ffn between different atom tasks and bond tasks.
         """
         super(FFNAtten, self).__init__()
 
         if constraint:
-            self.ffn = DenseLayers(first_linear_dim=features_size,
-                                   hidden_size=hidden_size,
-                                   num_layers=num_layers,
-                                   output_size=hidden_size,
-                                   dropout=dropout,
-                                   activation=activation)
+            if shared_ffn:
+                self.ffn = ffn_base
```

Comment:
```suggestion
                self.ffn = nn.Sequential([
                    ffn_base,
                    DenseLayers(
                        first_linear_dim=features_size,
                        hidden_size=hidden_size,
                        num_layers=1,
                        output_size=hidden_size,
                        dropout=dropout,
                        activation=activation
                    )
                ])
```

Reply:
I don't understand this suggestion, why would we need another layer here? ffn_base as currently constituted uses all but the last layer of the ffn, which is then added as the readout layer. I think actually my problem might be that the else statement in L139 uses the full number of layers and not num_layers - 1.